### PR TITLE
Adopt not discriminated Cadl-Ranch spec

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/microsoft/typespec/issues/4004

Also fixes an issue in the ordering of constructor parameters to order starting with the base model properties.

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/Providers/ModelFactories/ModelFactoriesCustomizationTests.cs [hunk 1]
```diff
@@ -135,7 +137,9 @@ public async Task CanCustomizeModelFullConstructor()
                             InputFactory.Property("Prop1", InputPrimitiveType.String, isRequired: true),
                         ])
                ],
-               compilation: async () => await Helpers.GetCompilationFromDirectoryAsync());
+               compilation: async () => await Helpers.GetCompilationFromDirectoryAsync(),
+               additionalMetadataReferences: [MetadataReference.CreateFromFile(typeof(BinaryData).Assembly.Location)]);
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/Providers/ModelFactories/ModelFactoriesCustomizationTests.cs:141]: why is this needed?

**Review by jorgerangel-msft — APPROVED** [review]

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/Providers/ModelFactories/ModelFactoriesCustomizationTests.cs (hunk 1 above)
Comment (JoshLove-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/Providers/ModelFactories/ModelFactoriesCustomizationTests.cs:141]: This was needed because otherwise the BinaryData symbols are not loaded correctly. It isn't needed for tests in MGC.CM because the ClientModelPlugin already adds the BinaryData reference for us.

Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/Providers/ModelFactories/ModelFactoriesCustomizationTests.cs:141]: I see, makes sense. thanks.
